src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -544,24 +557,132 @@ void MarkCustomAttributes (ICustomAttributeProvider provider, in DependencyInfo
 				MarkCustomAttribute (ca, reason);
 				MarkSpecialCustomAttributeDependencies (ca, provider);
 			}
+
+			// avoid allocating linker attributes for members that don't need them
```

Comment:
1. I assume we do this because otherwise we would populate the linker attribute cache for all methods/fields everywhere. If that's the case please change the comment to note about that. Also note that the existing usage for `DynamicallyAccessedMembersAttribute` will already populate the cache for all marked methods and fields - so I don't think we need this optimization here (otherwise we would have to add it to that case as well, but there it's probably a regression). I would still keep it in the assembly resolution step as that runs on everything, not just marked members.
1. I think this is worth adding as a functionality into the cache somehow - maybe we could simply have a `GetLinkerAttributesNoEmptyCache` or something like that? Which would act just like GetLinkerAttributes but it would not add a record for cases where there are no attributes to remember. The point being that you already have this same "logic/code" in two places.

Reply:
Originally I wanted to avoid setting up the cache for members without DynamicDependencyAttribute. `GetLinkerAttributesNoEmptyCache` would avoid inserting a struct into the Annotations dictionary for every member - on the other hand, it would still do unnecessary processing of linker attributes for unmarked members, and also requires members without linker attributes to be re-scanned later. Without something to measure it's not clear to me what is better.

For now, maybe we should just avoid populating the cache for members without any attributes. There will still be entries in the Annotations dictionary for members with non-linker attributes, but prevents a lot of the unnecessary structs already and doesn't require the callsite to use a different helper. See here: https://github.com/mono/linker/pull/1215/commits/5c637fad798051546245372ac7eb6c403f0d68c5#diff-efcf63149081c7bcad6e9ab7c490ac0bR433. If we hit a case where the perf is still an issue it might be easier to optimize that when we can measure it. What do you think?

Reply:
I like that idea... (hopefully nullable attributes won't ruin it :-)) - we should definitely do that.